### ISSUE TITLE
launch.script: identity: jarfolder might be ambiguous

### ISSUE THREAD

**Issue by uweschaefer** [spring-projects/spring-boot#5679]
We're using packaged jars to deploy our applications and run them in "service"-mode.

The identity, that is used as a part of the PIDs filename is (in our case) created in line 67 from the jarfolder. The jarfolder is defined as: `jarfolder="$(dirname "$jarfile")`

This turns out to be problematic, when the path contains links. I'll add a PR to add a `realpath` there.

**Comment by wilkinsona** [comment]
With reference to my comment on #5680, `readlink -f` seems to be a more widely available alternative to `realpath`

**Comment by uweschaefer** [comment]
Thanks for pointing the direction. I added a new PR.

**Comment by wilkinsona** [comment]
@uweschaefer  Looking at this more closely, I realise that I don't understand the problem that you're trying to solve. The script already follows symlinks to find the real jar, i.e. `jarfile` won't have any links in its path:

```
while [[ -L "$jarfile" ]]; do
  [[ "$jarfile" =~ init\.d ]] && init_script=$(basename "$jarfile")
  jarfile=$(readlink "$jarfile")
  cd "$(dirname "$jarfile")" || exit 1
  jarfile=$(pwd)/$(basename "$jarfile")
done
```

Can you please describe the problem that you have when the path contains links?

**Comment by alcCapone** [comment]
Hi,
you are right. The script resolves links already, but only if the jar file itself is a link. If however the path contains a segment which is a link the `$identity` turns out to be ambiguous. 

Example:

    ~/tmp$ ls foo
    demo-0.0.1-SNAPSHOT.jar
    
    ~/tmp$ ln -s foo bar

    ~/tmp$ ls bar
    demo-0.0.1-SNAPSHOT.jar

    ~/tmp$ MODE="service" PID_FOLDER=/tmp LOG_FOLDER=/tmp foo/demo-0.0.1-SNAPSHOT.jar start
    ~/tmp$ ls /tmp/*.pid
    /tmp/demo-0.0.1-SNAPSHOT_homecftmpfoo.pid

    ~/tmp$ MODE="service" PID_FOLDER=/tmp LOG_FOLDER=/tmp bar/demo-0.0.1-SNAPSHOT.jar start
    ~/tmp$ ls -1 /tmp/*.pid
    /tmp/demo-0.0.1-SNAPSHOT_homecftmpbar.pid
    /tmp/demo-0.0.1-SNAPSHOT_homecftmpfoo.pid

As a consequence we have to canonicalize the `$jarfolder` and resolve any link segment. In an OS-agnostic way we cannot use `readlink -f` because it doesn't work on OS X. Hence we have to come up with a work-around using `pwd -P`. I'll attach a PR.

**Comment by wilkinsona** [comment]
> we cannot use readlink -f because it doesn't work on OS X

The script targets Linux; working on OS X isn't important.

**Comment by wilkinsona** [comment]
Thanks for providing some more details. I'm not yet convinced that we should change this behaviour, and I'm especially cautious of doing so in 1.3.x as it could be a breaking change for some.

**Comment by alcCapone** [comment]
Hi, the serious problem that we're facing is that we 
- accidentally started the same services multiple times 
- weren't able to stop a service
- received incorrect statuses when asking if a service is running or stopped

depending on which path to the jar was used. BTW: we can also provide a PR using `readlink -f` instead.

**Comment by wilkinsona** [comment]
This is a great example of why it's a good idea to describe the problem rather than jumping straight in with a solution.

Can you please provide some more details about:

> - weren't able to stop a service
> - received incorrect statuses when asking if a service is running or stopped

**Comment by alcCapone** [comment]
I guess you're quite right!

I'd like to refer to my "foo" and "bar" example.
- I guess it's obvious that I'm able to start that same service twice
- If I start the service using the "foo" path I cannot stop it via the "bar" path.
- If I start the service using the "foo" path and I ask if the service is running via "bar" path I get a wrong status.

**Comment by wilkinsona** [comment]
Thanks. None of that behaviour seems unreasonable to me, and it's entirely possible that someone is relying upon it. If anything, I'm becoming convinced that we shouldn't change the current behaviour. Let's see what the rest of the team think.

**Comment by alcCapone** [comment]
Hi, I'd love to understand your perspective a bit better. Maybe I'm missing something here.

Let's assume we have a service that uses port (8080) and (if I don't resolve link segments in paths) have multiple paths to it. What sense does it make to be able to start the service multiple times? Necessary resources (8080) are blocked by the process started first.

**Comment by wilkinsona** [comment]
Someone may be using symlinks to ensure that the binary is identical but that the pid file and logs will be written to different location. They could then start the same binary multiple times with different configuration, for example by using a `SERVER_PORT` environment variable.

**Comment by alcCapone** [comment]
I get your point.

Just to describe ours: we have a path to a service application (`$BASEDIR/releases/app-x/current/app-x.jar`) and some kind of shortcut to it (`$BASEDIR/app-x -> $BASEDIR/releases/app-x/current`). So somebody could either use
- `$BASEDIR/releases/app-x/current/app-x.jar` or
- `$BASEDIR/app-x/app-x.jar`

Same service via two different paths is treated as two different services and hence
- can be started twice
- cannot stop service using path 1 if service has been started using path 2
- cannot find out if service is running using path 1 if service has been started using path 2

**Comment by uweschaefer** [comment]
Interesting. i definitely learned to discuss 'perceived' bugs before tying to provide a 'fix', because i could be missing some use-cases.

However, i would expected someone, who willingly starts several instances with different configurations to provider different APP_NAMEs to control the pid/log-filenames.
The fact the the variable involved is called 'identity' made me think it should be unambiguous.

**Comment by wilkinsona** [comment]
We decided to make the change that you have proposed, but to only do so in 1.4. That way we avoid breaking anyone who's relying on the current behaviour when they upgrade to a new maintenance release. It also gives us an opportunity to describe the behaviour change in the release notes.
